File: src/dotnet/PolyBus/Transport/Transactions/Messages/Handlers/Error/ErrorHandler.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace PolyBus.Transport.Transactions.Messages.Handlers.Error;

/// <summary>
/// A handler for processing message errors with retry logic.
/// </summary>
public class ErrorHandler
{
    /// <summary>
    /// The logger instance to use for logging.
    /// </summary>
    public ILogger<ErrorHandler> Log { get; set; } = NullLogger<ErrorHandler>.Instance;

    /// <summary>
    /// The delay increment in seconds for each delayed retry attempt.
    /// The delay is calculated as: delay attempt number * delay increment.
    /// </summary>
    public int DelayIncrement { get; set; } = 30;

    /// <summary>
    /// How many delayed retry attempts to make before sending to the dead-letter queue.
    /// </summary>
    public int DelayedRetryCount { get; set; } = 3;

    /// <summary>
    /// How many immediate retry attempts to make before applying delayed retries.
    /// </summary>
    public int ImmediateRetryCount { get; set; } = 3;

    /// <summary>
    /// The header key for storing error messages in dead-lettered messages.
    /// </summary>
    public string ErrorMessageHeader { get; set; } = "x-error-message";

    /// <summary>
    /// The header key for storing error stack traces in dead-lettered messages.
    /// </summary>
    public string ErrorStackTraceHeader { get; set; } = "x-error-stack-trace";

    /// <summary>
    /// The header key for storing the delayed retry count.
    /// </summary>
    public string RetryCountHeader { get; set; } = "x-retry-count";

    /// <summary>
    /// Retries the processing of a message according to the configured retry logic.
    /// </summary>
    public virtual async Task Retrier(IncomingTransaction transaction, Func<Task> next)
    {
        var delayedAttempt =
            transaction.IncomingMessage.Headers.TryGetValue(RetryCountHeader, out var headerValue)
            && byte.TryParse(headerValue, out var parsedHeaderValue)
            ? parsedHeaderValue
            : 0;
        var delayedRetryCount = Math.Max(1, DelayedRetryCount);
        var immediateRetryCount = Math.Max(1, ImmediateRetryCount);

        for (var immediateAttempt = 0; immediateAttempt < immediateRetryCount; immediateAttempt++)
        {
            try
            {
                await next();
                break;
            }
            catch (Exception error)
            {
                Log.LogError("Error processing message {MessageInfo} (immediate attempts: {immediateAttempt}, delayed attempts: {delayedAttempt}): {ErrorMessage}",
                    immediateAttempt,
                    delayedAttempt,
                    transaction.IncomingMessage.MessageInfo,
                    error.Message);

                transaction.OutgoingMessages.Clear();

                if (immediateAttempt < immediateRetryCount - 1)
                {
                    continue;
                }

                if (transaction.IncomingMessage.Bus.Transport.SupportsDelayedCommands
                    && delayedAttempt < delayedRetryCount)
                {
                    // Re-queue the message with a delay
                    delayedAttempt++;
                    var delayedMessage = new OutgoingMessage(
                        transaction.Bus,
                        transaction.IncomingMessage.Message,
                        transaction.Bus.Name,
                        transaction.IncomingMessage.MessageInfo)
                    {
                        DeliverAt = GetNextRetryTime(delayedAttempt),
                        Headers = transaction.IncomingMessage.Headers
                            .ToDictionary(it => it.Key, it => it.Value),
                    };
                    delayedMessage.Headers[RetryCountHeader] = delayedAttempt.ToString();
                    transaction.OutgoingMessages.Add(delayedMessage);

                    continue;
                }

                var deadLetterMessage = new OutgoingMessage(
                    transaction.Bus,
                    transaction.IncomingMessage.Message,
                    transaction.Bus.Transport.DeadLetterEndpoint,
                    transaction.IncomingMessage.MessageInfo)
                {
                    Headers = transaction.IncomingMessage.Headers
                        .ToDictionary(it => it.Key, it => it.Value),
                };
                deadLetterMessage.Headers[ErrorMessageHeader] = error.Message;
                deadLetterMessage.Headers[ErrorStackTraceHeader] = error.StackTrace ?? string.Empty;
                transaction.OutgoingMessages.Add(deadLetterMessage);
            }
        }
    }

    public virtual DateTime GetNextRetryTime(int attempt) => DateTime.UtcNow.AddSeconds(attempt * DelayIncrement);
}

File: src/kotlin/src/main/kotlin/polybus/transport/transactions/messages/handlers/error/ErrorHandler.kt
package polybus.transport.transactions.messages.handlers.error

import polybus.transport.transactions.IncomingTransaction
import polybus.transport.transactions.messages.OutgoingMessage
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.logging.Level
import java.util.logging.Logger

open class ErrorHandler {
    open var log: Logger = Logger.getLogger(ErrorHandler::class.java.name)
    open var delayIncrement: Int = 30
    open var delayedRetryCount: Int = 3
    open var immediateRetryCount: Int = 3
    open var errorMessageHeader: String = "x-error-message"
    open var errorStackTraceHeader: String = "x-error-stack-trace"
    open var retryCountHeader: String = "x-retry-count"

    open suspend fun retrier(transaction: IncomingTransaction, next: suspend () -> Unit) {
        var delayedAttempt = transaction.incomingMessage.headers[retryCountHeader]?.toIntOrNull() ?: 0
        val delayedRetries = maxOf(1, delayedRetryCount)
        val immediateRetries = maxOf(1, immediateRetryCount)

        for (immediateAttempt in 0 until immediateRetries) {
            try {
                next()
                break
            } catch (error: Throwable) {
                log.log(
                    Level.SEVERE,
                    "Error processing message ${transaction.incomingMessage.messageInfo} (immediate attempts: $immediateAttempt, delayed attempts: $delayedAttempt): ${error.message}",
                    error
                )

                transaction.outgoingMessages.clear()

                if (immediateAttempt < immediateRetries - 1) {
                    continue
                }

                if (transaction.incomingMessage.bus.transport.supportsDelayedCommands && delayedAttempt < delayedRetries) {
                    delayedAttempt += 1
                    val delayedMessage = OutgoingMessage(
                        transaction.bus,
                        transaction.incomingMessage.message,
                        transaction.bus.name,
                        transaction.incomingMessage.messageInfo
                    )
                    delayedMessage.deliverAt = getNextRetryTime(delayedAttempt)
                    delayedMessage.headers = transaction.incomingMessage.headers.toMutableMap()
                    delayedMessage.headers[retryCountHeader] = delayedAttempt.toString()
                    transaction.outgoingMessages.add(delayedMessage)
                    continue
                }

                val deadLetterMessage = OutgoingMessage(
                    transaction.bus,
                    transaction.incomingMessage.message,
                    transaction.bus.transport.deadLetterEndpoint,
                    transaction.incomingMessage.messageInfo
                )
                deadLetterMessage.headers = transaction.incomingMessage.headers.toMutableMap()
                deadLetterMessage.headers[errorMessageHeader] = error.message.orEmpty()
                deadLetterMessage.headers[errorStackTraceHeader] =
                    if (error.stackTrace.isEmpty()) "" else error.stackTraceToString()
                transaction.outgoingMessages.add(deadLetterMessage)
            }
        }
    }

    open fun getNextRetryTime(attempt: Int): Instant =
        Instant.now().plus((attempt * delayIncrement).toLong(), ChronoUnit.SECONDS)
}
